overlord/snapstate/snapstate.go

```diff
@@ -1969,17 +1969,132 @@ func updateManyFiltered(ctx context.Context, st *state.State, names []string, re
 		return nil, nil, err
 	}
 
-	updated, updateTss, err := doUpdate(ctx, st, names, toUpdate, params, userID, flags, nil, deviceCtx, fromChange)
-	if err != nil {
-		return nil, nil, err
+	if deviceCtx.IsCoreBoot() && release.OnClassic {
+		essential, nonEssential := splitEssentialUpdates(deviceCtx, toUpdate, names)
+		var essentialTss, nonEssentialTss []*state.TaskSet
+		var preDlTss []*state.TaskSet
+		var updatedAll []string
+
+		// taskset with essential snaps (snapd, kernel, gadget and their bases)
+		updated, tss, err := doUpdate(ctx, st, essential.names, essential.infos, params, userID, flags, nil, deviceCtx, fromChange)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if len(tss.PreDownload) > 0 {
+			preDlTss = append(preDlTss, tss.PreDownload...)
+		}
+
+		// if there are only pre-downloads, don't add a check-rerefresh task
+		if len(tss.Refresh) > 0 {
+			tss.Refresh = finalizeUpdate(st, tss.Refresh, len(updates) > 0, updated, userID, flags)
+			essentialTss = tss.Refresh
+			updatedAll = append(updatedAll, updated...)
+		}
+
+		// taskset with non-essential snaps (apps and their bases)
+		updated, tss, err = doUpdate(ctx, st, nonEssential.names, nonEssential.infos, params, userID, flags, nil, deviceCtx, fromChange)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if len(tss.PreDownload) > 0 {
+			preDlTss = append(preDlTss, tss.PreDownload...)
+		}
+
+		// if there are only pre-downloads, don't add a check-rerefresh task
+		if len(tss.Refresh) > 0 {
+			tss.Refresh = finalizeUpdate(st, tss.Refresh, len(updates) > 0, updated, userID, flags)
+			nonEssentialTss = tss.Refresh
+			updatedAll = append(updatedAll, updated...)
+		}
+
+		if len(essentialTss) > 0 && len(nonEssentialTss) > 0 {
+			// if snapd is in the essential snaps set, the non-essentials must wait for it
+			if strutil.ListContains(updatedAll, "snapd") {
+				snapdTss, err := findTasksetForSnap(essentialTss, "snapd")
+				if err != nil {
+					return nil, nil, err
+				}
+
+				// make non-essential snaps also wait for snapd
+				for _, ts := range nonEssentialTss {
+					ts.WaitAll(snapdTss)
```

Comment:
I think we switched to try to avoid using WaitAll across many to many task, Philip had to deal with this, ask him

Reply:
I've changed this to add a single `WaitFor` between each app's first task and snapd's last (same with the bases). Thank for the feedback